plugins/repository-gcs/src/main/java/io/crate/gcs/GCSBlobStore.java

```diff
@@ -55,10 +102,254 @@ public BlobContainer blobContainer(BlobPath path) {
+                throw se;
+            }
+        }
+        throw storageException;
+    }
+
+    /**
+     * Uploads a blob using the "multipart upload" method (a single
+     * 'multipart/related' request containing both data and metadata. The request is
+     * gziped), see:
+     * https://cloud.google.com/storage/docs/json_api/v1/how-tos/multipart-upload
+     *
+     * @param blobInfo            the info for the blob to be uploaded
+     * @param inputStream         the stream containing the blob data
+     * @param blobSize            the size
+     * @param failIfAlreadyExists whether to throw a FileAlreadyExistsException if the given blob already exists
+     */
+    private void writeBlobMultipart(BlobInfo blobInfo, InputStream inputStream, long blobSize, boolean failIfAlreadyExists)
+        throws IOException {
+        assert blobSize <= getLargeBlobThresholdInBytes() : "large blob uploads should use the resumable upload method";
+        byte[] buffer = new byte[Math.toIntExact(blobSize)];
+        inputStream.readNBytes(buffer, 0, buffer.length);
         try {
-            storage.close();
-        } catch (Exception e) {
-            throw new IOException(e);
+            final Storage.BlobTargetOption[] targetOptions = failIfAlreadyExists
+                ? new Storage.BlobTargetOption[]{Storage.BlobTargetOption.doesNotExist()}
+                : new Storage.BlobTargetOption[0];
+            client().create(blobInfo, buffer, targetOptions);
+        } catch (StorageException se) {
+            if (failIfAlreadyExists && se.getCode() == HTTP_PRECON_FAILED) {
+                throw new FileAlreadyExistsException(blobInfo.getBlobId().getName(), null, se.getMessage());
+            }
+            throw se;
         }
     }
+
+    void deleteDirectory(String pathStr) throws IOException {
+        Page<Blob> page = client().list(bucketName, BlobListOption.prefix(pathStr));
+        do {
+            Collection<String> blobsToDelete = new ArrayList<>();
+            page.getValues().forEach(b -> {
+                blobsToDelete.add(b.getName());
+            });
+            deleteBlobsIgnoringIfNotExists(blobsToDelete);
+            page = page.getNextPage();
+        } while (page != null);
+    }
+
+    void deleteBlobsIgnoringIfNotExists(Collection<String> blobNames) throws IOException {
+        if (blobNames.isEmpty()) {
+            return;
+        }
+        var blobIdsToDelete = new ArrayList<BlobId>();
+        for (String blobName : blobNames) {
+            blobIdsToDelete.add(BlobId.of(bucketName, blobName));
+        }
+        List<BlobId> failedBlobs = Collections.synchronizedList(new ArrayList<>());
+        try {
+            final AtomicReference<StorageException> ioe = new AtomicReference<>();
+            final StorageBatch batch = client().batch();
+            for (BlobId blob : blobIdsToDelete) {
+                batch.delete(blob).notify(new BatchResult.Callback<Boolean, StorageException>() {
```

Comment:
```suggestion
    void deleteDirectory(String pathStr) throws IOException {
        Page<Blob> page = client().list(bucketName, BlobListOption.prefix(pathStr));
        do {
            deleteBlobsIgnoringIfNotExists(Iterables.transform(page.getValues(), Blob::getName));
            page = page.getNextPage();
        } while (page != null);
    }

    void deleteBlobsIgnoringIfNotExists(Iterable<String> blobNames) throws IOException {
        Iterator<String> blobNamesIt = blobNames.iterator();
        if (!blobNamesIt.hasNext()) {
            return;
        }
        List<BlobId> failedBlobs = Collections.synchronizedList(new ArrayList<>());
        try {
            final AtomicReference<StorageException> ioe = new AtomicReference<>();
            final StorageBatch batch = client().batch();
            for (String blobName : blobNames) {
                BlobId blob = BlobId.of(bucketName, blobName);
                batch.delete(blob).notify(new BatchResult.Callback<Boolean, StorageException>() {
```

Would avoid a couple of copy operations that seem unnecessary(?)